lib/semantics/check-omp-structure.h

```diff
@@ -0,0 +1,148 @@
+  void Enter(const parser::OmpClause::Copyin &);
+  void Enter(const parser::OmpClause::Copyprivate &);
+  void Enter(const parser::OmpClause::Device &);
+  void Enter(const parser::OmpClause::DistSchedule &);
+  void Enter(const parser::OmpClause::Final &);
+  void Enter(const parser::OmpClause::Firstprivate &);
+  void Enter(const parser::OmpClause::From &);
+  void Enter(const parser::OmpClause::Grainsize &);
+  void Enter(const parser::OmpClause::Lastprivate &);
+  void Enter(const parser::OmpClause::NumTasks &);
+  void Enter(const parser::OmpClause::NumTeams &);
+  void Enter(const parser::OmpClause::NumThreads &);
+  void Enter(const parser::OmpClause::Ordered &);
+  void Enter(const parser::OmpClause::Priority &);
+  void Enter(const parser::OmpClause::Private &);
+  void Enter(const parser::OmpClause::Safelen &);
+  void Enter(const parser::OmpClause::Shared &);
+  void Enter(const parser::OmpClause::Simdlen &);
+  void Enter(const parser::OmpClause::ThreadLimit &);
+  void Enter(const parser::OmpClause::To &);
+  // TODO: Link
+  void Enter(const parser::OmpClause::Uniform &);
+  void Enter(const parser::OmpClause::UseDevicePtr &);
+  // TODO: IsDevicePtr
+
+  void Enter(const parser::OmpAlignedClause &);
+  void Enter(const parser::OmpDefaultClause &);
+  void Enter(const parser::OmpDependClause &);
+  void Enter(const parser::OmpIfClause &);
+  void Enter(const parser::OmpLinearClause &);
+  void Enter(const parser::OmpMapClause &);
+  void Enter(const parser::OmpProcBindClause &);
+  void Enter(const parser::OmpReductionClause &);
+  void Enter(const parser::OmpScheduleClause &);
+
+  bool HasInvalidCloselyNestedRegion(
+      const parser::CharBlock &, const OmpDirectiveEnumType &);
+  void CheckAllowed(const OmpClauseEnum &);
+
+private:
+  struct OmpContext {
+    parser::CharBlock currentDirectiveSource{nullptr};
+    parser::CharBlock currentClauseSource{nullptr};
+    OmpDirectiveEnum currentDirectiveEnum;
+    OmpClauseEnumType allowedClauses;
+    OmpClauseEnumType allowedOnceClauses;
+    OmpClauseEnumType seenClauses;
+  };
+  const OmpContext &GetOmpContext() const { return ompContext_.back(); }
+  void SetOmpContextDirectiveSource(const parser::CharBlock &directive) {
+    ompContext_.back().currentDirectiveSource = directive;
+  }
+  void SetOmpContextClauseSource(const parser::CharBlock &clause) {
+    ompContext_.back().currentClauseSource = clause;
+  }
+  void SetOmpContextDirectiveEnum(const OmpDirectiveEnum &dir) {
+    ompContext_.back().currentDirectiveEnum = dir;
+  }
+  void SetOmpContextAllowed(const OmpClauseEnumType &allowed) {
+    ompContext_.back().allowedClauses = allowed;
+  }
+  void SetOmpContextAllowedOnce(const OmpClauseEnumType &allowedOnce) {
+    ompContext_.back().allowedOnceClauses = allowedOnce;
+  }
+  void SetOmpContextSeen(const OmpClauseEnum &seenType) {
+    ompContext_.back().seenClauses.set(seenType);
+  }
+
+  SemanticsContext &context_;
+  std::vector<OmpContext> ompContext_;
```

Comment:
Can you indicate with comment that this is being used as a stack and that ompContext_.back() is the top of the stack ?

Reply:
Okay I will.

Reply:
Addressed in commit [cb89744](cb89744746186bcf120d8ca62a791e2778366b9b).